Save credentials and add remember me option

Shubhamr837 opened the pull request.

linl33 commented:
runtests

linl33 commented:
@Shubhamr837 I will review this PR this weekend. I look a quick look and noticed you are storing the user provided information in a properties file. You might want to look into the [`Java Preferences API`](https://docs.oracle.com/javase/8/docs/technotes/guides/preferences/overview.html) so we don't need to manage the storage ourselves.

linl33 commented on src/main/java/org/opendatakit/suitcase/net/SyncWrapper.java:
```diff
@@ -60,8 +60,6 @@ public void init(CloudEndpointInfo cloudEndpointInfo) throws IOException, JSONEx
 
       this.sc.init(host, this.cloudEndpointInfo.getUserName(), this.cloudEndpointInfo.getPassword());
-      
-      updateTableList();
```
Why was this removed?

Shubhamr837 replied:
I have removed it from here to login task. After syncWrapper is initialized we can call update table list from login task and login task will also update the status is the progress bar. This is done because after login is done from saved credentials table id list will have to be updated and user needs to know that table list is being updated as initially the table list will be blank. If saved credentials are incorrect the method would through exception, we can handle the error in login task and send user back to login page.

linl33 commented on src/main/java/org/opendatakit/suitcase/ui/PullPanel.java:
```diff
@@ -65,11 +72,22 @@ public PullPanel(IOPanel parent) {
+            }
+        });
+        gbc.weighty = 1;
+        this.add(logoutButton);
```
The logout button construction should be its own method, so it's easier to read.

linl33 commented on src/main/java/org/opendatakit/suitcase/ui/PullPanel.java:
```diff
@@ -138,17 +156,42 @@ public void actionPerformed(ActionEvent e) {
-            sPullButton.setText(DOWNLOAD_LABEL);
-            parent.setButtonsState(ButtonState.ENABLED, ButtonState.ENABLED, ButtonState.ENABLED,ButtonState.ENABLED);
+        if (evt.getNewValue() != null) {
+            switch (evt.getPropertyName()) {
```
Possible NPE here. Check `PropertyChangeEvent` javadoc.

linl33 reviewed (changes requested):
Left a few minor comments, let me know if you have questions. We also discussed quite a few changes on Friday's call. Please re-request review when these are done. Thanks for working on this.

Shubhamr837 commented on src/main/java/org/opendatakit/suitcase/net/LoginTask.java:
```diff
@@ -27,8 +30,12 @@ protected Void doInBackground() throws IOException, JSONException {
     syncWrapper.init(cloudEndpointInfo);
-    syncWrapper.setPrivilegesInfo();
-
+    if(syncWrapper.isInitialized()) {
```
after the init method is called it will be initialized again https://github.com/Shubhamr837/suitcase/blob/c2461b8b74378d34cb189e5f1208eecf47df9188/src/main/java/org/opendatakit/suitcase/net/SyncWrapper.java#L64

Shubhamr837 commented:
Thanks for the review @linl33 I have implemented the changes as discussed in the last meeting and replied the the comments I couldn't resolve. Please Review

codeknight3 commented on src/main/java/org/opendatakit/suitcase/model/SyncClientError.java:
```diff
@@ -0,0 +1,28 @@
+        String errorMsg = e.getCause().getMessage();
+        int startIndex = errorMsg.indexOf("Status");
+        if (startIndex != -1) {
+            statusCode = Integer.parseInt(errorMsg.substring(startIndex + 7, startIndex + 10));
```
You can add comments about the use of startIndex + 7, startIndex + 13, or one sample error message. So that it will be easier in the future to understand code.

codeknight3 commented on src/main/java/org/opendatakit/suitcase/net/LoginTask.java:
```diff
@@ -56,7 +57,13 @@ protected void finished() {
       if (cause instanceof JSONException) {
-        errMsg = BAD_CRED;
+        SyncClientError syncClientError = new SyncClientError(e);
+        if(syncClientError.getStatusCode()==401){
```
Add space. (syncClientError.getStatusCode() == 401)

codeknight3 commented on src/main/java/org/opendatakit/suitcase/ui/PullPanel.java:
```diff
@@ -138,17 +159,42 @@ public void actionPerformed(ActionEvent e) {
+            Properties appProperties = new Properties();
+            try (FileInputStream fileInputStream = new FileInputStream(SuitcaseConst.PROPERTIES_FILE)) {
+                appProperties.load(fileInputStream);
+                appProperties.clear();
```
Why deleting stored data after logout? If the user logouts and closes the application after that when the user tries to log in again we still want its data to be there. Right?

Shubhamr837 replied:
Yes sir I am modifying it now. Only password will be removed.

codeknight3 replied:
Yes. So I think correct behavior should be like this:
If the save password box is unchecked - remove the password after logout
If the save password box is checked - No need to remove anything

Shubhamr837 replied:
Sir the password is not saved in the properties file after login if save password is unchecked. So one way is: 
 1. checking that password field is there or not  
 2.  Remove if it is there.
 If any other implementation would be more efficient please suggest.

codeknight3 replied:
So in that case, you don't need to change the properties file at all after the user logout. 
If the user checked the save password box password will be stored in the properties file and we can reuse it when the user restarts the application.
In case of save password box is unchecked, password will not be stored in the properties file which is what we want.

Shubhamr837 replied:
Yes sir but this will work but only if user stays in the application after logout. Because after user opens the application again auto login logic is implemented in this way:
1. If password field is present login the user
2. Else if password field is absent only fill the server_url, app_id and username and leave password text field blank.

Shubhamr837 replied:
So after we remove password it will not login the user when application is opened again.

codeknight3 replied:
But we want a password to be filled in automatically when the user opens the application again and he previously had checked the save password box.  Right?
A simple solution which I can think of,

Start of application: If the properties file is present fill all the fields using values in the properties file, else create properties file with all the keys, and values set to an empty string. Run Auto login logic only when password is not an empty string. 
After Successful login: Change all the values in the properties file corresponding to user-provided values. If the save password box is not checked set the value of the password(in the properties file) to an empty string.
After Logout: No need to change anything in the properties file.

It's just a suggestion. If you think there is some easier way, you can implement that.

Shubhamr837 commented:
@linl33 @codeknight3 I have replaced properties file with preferences. It has made the code simpler and short.
For html parsing I have also added comments that make it easy to understand.
I have also sent a PR for sync client library that adds the html response to the exception message.
https://github.com/odk-x/sync-client/pull/47
